source/server-discovery-and-monitoring/tests/spec/isMaster-timeout.yml

```diff
@@ -0,0 +1,225 @@
+      # Unknown and clear the pool.
+      - name: waitForEvent
+        object: testRunner
+        arguments:
+          event: ServerMarkedUnknownEvent
+          count: 1
+      - name: waitForEvent
+        object: testRunner
+        arguments:
+          event: PoolClearedEvent
+          count: 1
+      # Perform an operation to ensure the node is discovered.
+      - name: insertMany
+        object: collection
+        arguments:
+          documents:
+            - _id: 1
+            - _id: 2
+      # We cannot assert the server was marked Unknown and pool was cleared an
+      # exact number of times because the RTT isMaster may or may not have
+      # triggered this failpoint as well.
+      # - name: assertEventCount
+      #   object: testRunner
+      #   arguments:
+      #     event: ServerMarkedUnknownEvent
+      #     count: 1
+      # - name: assertEventCount
+      #   object: testRunner
+      #   arguments:
+      #     event: PoolClearedEvent
+      #     count: 1
+
+    expectations:
+      - command_started_event:
+          command:
+            insert: *collection_name
+            documents:
+              - _id: 1
+              - _id: 2
+          command_name: insert
+          database_name: *database_name
+
+    outcome:
+      collection:
+        data:
+          - {_id: 1}
+          - {_id: 2}
+
+  - description: Network timeout on Monitor check
+    clientOptions:
+      retryWrites: false
+      connectTimeoutMS: 750
+      heartbeatFrequencyMS: 500
+      appname: timeoutMonitorCheckTest
+    operations:
+      # Perform an operation to ensure the node is discovered.
+      - name: insertMany
+        object: collection
+        arguments:
+          documents:
+            - _id: 1
+            - _id: 2
+      # Configure the next streaming isMaster check to fail with a timeout
+      # Use times: 2 so that the RTT isMaster is blocked as well.
+      - name: configureFailPoint
+        object: testRunner
+        arguments:
+          failPoint:
+            configureFailPoint: failCommand
+            mode: { times: 2 }
+            data:
+                failCommands: ["isMaster"]
+                appName: timeoutMonitorCheckTest
+                blockConnection: true
+                blockTimeMS: 1000
+      # The network error on the next check should mark the server Unknown and
+      # clear the pool.
+      - name: waitForEvent
+        object: testRunner
+        arguments:
+          event: ServerMarkedUnknownEvent
+          count: 1
+      - name: waitForEvent
+        object: testRunner
+        arguments:
+          event: PoolClearedEvent
+          count: 1
+      # Perform an operation to ensure the node is rediscovered.
+      - name: insertMany
+        object: collection
+        arguments:
+          documents:
+            - _id: 3
+            - _id: 4
+      # Assert the server was marked Unknown and pool was cleared exactly once.
```

Comment:
I think there is ambiguity here with the new spec prose. [network error on a server check](https://github.com/mongodb/specifications/blob/16894a602b8ff7ddf33785f36700d34275f5d84e/source/server-discovery-and-monitoring/server-monitoring.rst#network-error-during-server-check) now says "When a server check fails due to a network error (including a network timeout), the client MUST mark the server Unknown and clear the connection pool for the server. If the monitor's socket is bad it is likely that all are" and yet this test is verifying that the pool is cleared exactly once. The linked prose further links to "retry ismaster calls once" but the pseudocode doesn't do this retry, and I can't find text describing that retry either.

Reply:
I _think_ this test is consistent with the "the client MUST mark the server Unknown and clear the connection pool for the server" rule. Here is what I expect to happen starting from the configureFailPoint:

1. testRunner configures the server to block the next two isMaster commands for 1000ms each.
1. We guarantee that both the streaming connection and the RTT connection will hit this failpoint within 500ms because `heartbeatFrequencyMS=500` and `blockTimeMS: 1000`.
1. Both the streaming connection and the RTT connection will receive timeouts because the `blockTimeMS` > `connectTimeoutMS`.
1. The RTT timeout error does not mark the server Unknown.
1. The streaming timeout error  marks the server Unknown, clears the pool, and runs an immediate check (without sleeping).